Source/PeleLMeX_Forces.cpp

```diff
@@ -47,201 +49,180 @@ PeleLM::getVelForces(
-    const auto& temp_arr = (m_incompressible) != 0
-                             ? DummyFab.array()
-                             : ldata_p->state.const_array(mfi, TEMP);
-    const auto& extmom_arr = m_extSource[lev]->const_array(mfi, VELX);
-    const auto& extrho_arr = m_extSource[lev]->const_array(mfi, DENSITY);
-    const auto& force_arr = a_velForce->array(mfi);
-
-    // Get other forces (gravity, ...)
-    getVelForces(
-      lev, bx, time, force_arr, vel_arr, rho_arr, rhoY_arr, rhoh_arr, temp_arr,
-      extmom_arr, extrho_arr);
+  const int pseudo_gravity = m_ctrl_pseudoGravity;
+  const amrex::Real dV_control = m_ctrl_dV;
+  const auto grav = m_gravity;
+  const auto gp0 = m_background_gp;
+  const int ps_dir = m_ctrl_flameDir;
+
+  auto const& state_ma = ldata_p->state.const_arrays();
+  auto const& ext_ma = m_extSource[lev]->const_arrays();
+  auto const& force_ma = a_velForce->arrays();
+  auto const& gp_ma = ldataGP_p->gp.const_arrays();
+  auto const& divTau_ma =
+    (has_divTau != 0) ? a_divTau->const_arrays() : state_ma;
 
 #ifdef PELE_USE_PLASMA
-    const auto& phiV_arr = ldata_p->state.const_array(mfi, PHIV);
-    const auto& ne_arr = ldata_p->state.const_array(mfi, NE);
-    addLorentzVelForces(lev, bx, time, force_arr, rhoY_arr, phiV_arr, ne_arr);
+  auto const& dx = geom[lev].CellSizeArray();
+  auto const& ba = grids[lev];
+  auto const& zkl = zk;
 #endif
 
-    // Add forcing terms to maintain turbulence
-    if (m_do_turbulent_forcing) {
-      // note: if m_incompressible == false then m_rho is unused by
-      // addTurbVelForces
+  amrex::ParallelFor(
+    *a_velForce,
+    [state_ma, ext_ma, force_ma, grav, gp0, ps_dir, add_gradP, gp_ma,
+     has_divTau, divTau_ma, is_incomp, rho_incomp, pseudo_gravity, dV_control
+#ifdef PELE_USE_PLASMA
+     ,
+     dx, time, ba, zkl
+#endif
+  ] AMREX_GPU_DEVICE(int box_no, int i, int j, int k) {
+      amrex::Array4<amrex::Real const> vel(state_ma[box_no], VELX);
+      amrex::Array4<amrex::Real const> extmom(ext_ma[box_no], VELX);
+      amrex::Array4<amrex::Real const> rho(state_ma[box_no], DENSITY);
+      amrex::Array4<amrex::Real const> extrho(ext_ma[box_no], DENSITY);
+      // overwrite if incompressible, just to point to something
+      if (is_incomp != 0) {
+        rho = vel;
+        extrho = extmom;
+      }
+      // background gp, pseudo grav, ext sources
+      makeVelForce(
+        i, j, k, is_incomp, rho_incomp, pseudo_gravity, ps_dir, grav, gp0,
+        dV_control, vel, rho, extmom, extrho, force_ma[box_no]);
+#ifdef PELE_USE_PLASMA
+      amrex::Array4<amrex::Real const> rhoY(state_ma[box_no], FIRSTSPEC);
+      amrex::Array4<amrex::Real const> rhoh(state_ma[box_no], RHOH);
+      amrex::Array4<amrex::Real const> temp(state_ma[box_no], TEMP);
+      amrex::Array4<amrex::Real const> phiV(state_ma[box_no], PHIV);
+      amrex::Array4<amrex::Real const> nE(state_ma[box_no], NE);
+      amrex::GpuArray<int, 3> blo = ba[box_no].loVect3d();
```

Comment:
BoxArray is not a device-safe class so this fails on GPU

Reply:
I think that should be ok. It compiles fine on GPU now at least. But it looks like another of the recent changes you made is causing the FlameSheetIons case to fail even on CPU.